examples/gif_paintable/main.rs

```diff
@@ -0,0 +1,239 @@
+                self.width.take(),
+                self.height.take(),
+            ) {
+                let w = width as f64;
+                let h = height as f64;
+                texture.snapshot(snapshot, w, h);
+            } else {
+                let snapshot = snapshot.downcast_ref::<gtk::Snapshot>().unwrap();
+                snapshot.append_color(
+                    &gdk::RGBA::black(),
+                    &graphene::Rect::new(0f32, 0f32, width as f32, height as f32),
+                );
+            }
+        }
+    }
+}
+
+glib::wrapper! {
+    pub struct GifPaintable(ObjectSubclass<imp::GifPaintable>) @implements gdk::Paintable;
+}
+
+impl GifPaintable {
+    pub fn new() -> Self {
+        glib::Object::new(&[]).expect("Failed to create object of type GifPaintable")
+    }
+
+    // Loads the bytes of a GIF into the paintable
+    // The loading consists of decoding the gif with a GIFDecoder
+    // Then storing the frames so that the paintable can render them
+    //
+    // You can only call this function once due to the callback system in setup_next_frame
+    // causing the frames to refresh at faster intervals
+    pub fn load_from_bytes(&self, bytes: &[u8]) -> Result<(), Box<dyn std::error::Error>> {
+        let self_ = imp::GifPaintable::from_instance(self);
+        self_.current_idx.set(0);
+
+        let read = Cursor::new(bytes);
+        let decoder = GifDecoder::new(read)?;
+        let frames = decoder
+            .into_frames()
+            .collect_frames()?
+            .into_iter()
+            .map(Frame::new)
+            .collect::<Vec<Frame>>();
+
+        self_.frames.replace(Some(frames));
+
+        // make sure the first frame is queued to play
+        self.setup_next_frame();
+
+        Ok(())
+    }
+
+    fn setup_next_frame(&self) {
+        let self_ = imp::GifPaintable::from_instance(self);
+        let idx = self_.current_idx.get();
+        let frames_ref = self_.frames.borrow();
+
+        // if we have stored no frames then we early return early
+        // and instead render a default frame in `imp::GifPaintable::snapshot`
+        let frames = match frames_ref.deref() {
+            Some(frames) => frames,
+            None => return
+        };
+
+        let next_frame = frames.get(idx).unwrap();
+
+        self_.width.set(Some(next_frame.width));
+        self_.height.set(Some(next_frame.height));
+        self_.next_frame.replace(Some(next_frame.texture.clone()));
+
+        // invalidate the contents so that the new frame will be rendered
+        self.invalidate_contents();
+
+        // setup a callback to this function once the frame has finished so that
+        // we can play the next frame
+        let update_next_frame_callback = glib::clone!(@weak self as paintable => move || {
+            paintable.setup_next_frame();
+        });
+
+        glib::timeout_add_local_once(next_frame.frame_duration, update_next_frame_callback);
```

Comment:
I saw this PR by coincidence, and I've implemented something similar myself.

Using frame_duration directly as the delay here will cause the gif to drift slightly behind its full speed, and that'll be worse if any work is being done on the main thread that can delay the callback even slightly. The time that this frame was intended to be displayed should be stored and compared against the current time to correct for drift when scheduling the next frame.

If the application does take too long between frames then you have to choose between slowing down the gif or skipping frames, but that's probably down to taste.